Add support for allowedShippingCountries in Express Checkout Element

asumaran opened the pull request:
Implements https://github.com/Automattic/woocommerce-payments/issues/9104

## Changes proposed in this Pull Request

Adds support for specifying allowed shipping countries (`allowedShippingCountries`) in the Express Checkout Element.

## Testing instructions

> [!TIP]  
> If you want to test with multiple addresses from different countries, you can add yourself to the Google test card suite group. In test mode, you’ll have access to a set of test cards, rather than the cards associated with your production account.
 https://developers.google.com/pay/api/web/guides/resources/test-card-suite

> [!NOTE]  
> To test the changes introduced here, we first need to create shipping options for at least two countries. In this case the countries will be USA and Brazil. You can create a "Flat rate" shipping method for both countries. That will be sufficient.

> [!NOTE]  
> The test results should be the same for all the pages affected by this PR: Product Page, Cart Page, Checkout Page, Cart shortcode, and Checkout shortcode. (Does not apply to the "Pay for Order" page.)

> [!NOTE]
> All cases should be tested in the same way:
>1. If you are on the page of a physical product (that is, it requires shipping), simply click the Express Checkout button and proceed to checkout.
>2. If you are testing on other pages (except "Pay for Order"), add a product to the cart and then visit the desired page. Click the Express Checkout button and proceed to checkout.
>3. **There's no need to test the "Pay for Order"** page since all orders on this page are defined [as not requiring shipping](https://github.com/Automattic/woocommerce-payments/blob/7f8cad70de8488f5bdac3599d21697b50b1e2416/client/express-checkout/index.js#L594) in the Express Checkout settings.

> [!NOTE]
> Since this PR includes many combinations to test, and given that ECE has two different implementations, I would suggest testing at least the Checkout (block) page and the Shortcode Checkout page. **That would cover both ECE implementations currently in WooPayments**.

> [!IMPORTANT]
> Test with Apple Pay on at least the pages mentioned above. The results should be the same for each test.

### Case 1: "Shipping location(s)" is set to "Ship to all countries you sell to"

- **"Selling location(s)" is set to "Sell to all countries"**
  - ✅ The checkout process should work without issues for any shipping address from any country.
  - Note that it won't be possible to checkout using a shipping address from a country without shipping options.
- **"Selling location(s)" is set to "Sell to all countries, except for…"** and "Sell to all countries, except for…" is set to "United Kingdom (UK)"
  - ✅ Checkout should work without issues for any shipping address except one from the UK.
  - Note that it won’t be possible to checkout using a shipping address from a country without shipping options. For this PR, it should be possible to checkout with shipping addresses from the USA and Brazil only, as these are the only countries with shipping options.
- **"Selling location(s)" is set to "Sell to specific countries"** and “Sell to specific countries” is set to "United States (US)"
  - ✅ Checkout should complete without issues with any US address.
  - It shouldn't be possible to select a shipping address from Brazil or any country other than the USA.

### Case 2: "Shipping location(s)" is set to "Ship to all countries"

This option ignores the value set in "Selling location(s)", therefore:

- ✅ Checkout should complete without issues for all countries.
- Note that it won't be possible to checkout using a shipping address from a country without shipping options.
- Therefore, for this case, it should allow shipping to the USA and Brazil. If desired, you can add another country with shipping options to verify that shipping works to that country as well.

### Case 3: "Shipping location(s)" is set to "Ship to specific countries only" and "Ship to specific countries" is set to "Brazil"

- ✅ Checkout should complete without issues with shipping addresses from Brazil.
- It shouldn’t be possible to select a shipping address from any other country.

### Potential Issues:

If, during testing, you find that you can’t open the payment sheet and see the following error in the browser console:


```
Uncaught (in promise) IntegrationError: When `shippingAddressRequired` is true, you must specify `shippingRates`.
```


This is a known issue and has been reported [here](https://github.com/Automattic/woocommerce-payments/issues/9593). One way to fix it is by clearing all cookies and starting again. Another way is to go to the checkout page and change the shipping address to a country with shipping options, such as the USA or Brazil.

-------------------

- [ ] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [ ] Covered with tests (or have a good reason not to test in description ☝️)
- [ ] Tested on mobile (or does not apply)

**Post merge**

<!--
Make sure you edit the page for the current release when adding testing instructions.
We often create a blank page ahead of time for the next release.
If this PR need not be QA tested, edit to 'QA Testing Not Applicable'
-->

- [ ] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : _Add link here / 'QA Testing Not Applicable'_
- [ ] Add or update [critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows) and [testing instructions for critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows), if applicable.
- [ ] Add what's changed (description, screenshot, demo videos etc.) to the release announcement post, if applicable.

asumaran committed:
Expose allowed shipping countries to the frontend

asumaran committed:
Define allowedShippingCountries on Stripe for Express Checkouts

botwoo commented:
### Test the build 

#### Option 1. Jetpack Beta

- Install and activate [Jetpack Beta](https://jetpack.com/download-jetpack-beta/).
- Use this build by searching for PR number `9592` or branch name `9104-add-support-for-allowedshippingcountries-in-express-checkout-element` in your-test.site/wp-admin/admin.php?page=jetpack-beta&plugin=woocommerce-payments          

#### Option 2. Jurassic Ninja - available for logged-in A12s

:rocket: [Launch a JN site with this branch](https://jurassic.ninja/create/?jetpack-beta&shortlived&nojetpack&woocommerce&woocommerce-payments-dev-tools&branches.woocommerce-payments=9104-add-support-for-allowedshippingcountries-in-express-checkout-element) :rocket:

:information_source: Install this [Tampermonkey script](https://github.com/Automattic/woocommerce-payments/tree/develop/bin/wcpay-live-branches) to get more options.

---

Build info: 

- Latest commit: dbbf82f0d6a11ceaddcefd14cf99a530c1e46d12
- Build time: 2024-10-28 22:04:42 UTC

_Note: the build is updated when a new commit is pushed to this PR._

asumaran committed:
Add changelog entry

asumaran committed:
Add allowedShippingCountries para to ECE in product and shortcode pages

asumaran committed:
Do not send list of countries if store ships to all countries

asumaran committed:
Skip sending full country list when unnecessary

Don’t send the full country list in the two cases where all countries would be sent

asumaran commented on includes/express-checkout/class-wc-payments-express-checkout-button-handler.php:
```diff
@@ -257,6 +257,15 @@ public function scripts() {
+		$ship_to_all = get_option( 'woocommerce_ship_to_countries' ) === 'all';
+		// 2. ships to all countries it sells to and store sells to all countries.
+		$ship_to_all_and_selling_to_all = get_option( 'woocommerce_allowed_countries' ) === 'all' && empty( get_option( 'woocommerce_ship_to_countries' ) );
+		if ( ! $ship_to_all && ! $ship_to_all_and_selling_to_all ) {
```
There's no real need to add a condition here. I'm only doing this to be extra cautious in case Stripe supports a country code that WooPayments (or rather WooCommerce) doesn't include in its list. For example, if Stripe supports the country code 'ZZ', but WooPayments doesn't provide "ZZ" on the country list, the best way to handle 'ZZ' is to send no country list and let Stripe handle all countries (which is the default behavior when no list is provided)

asumaran replied:
I wanted to double check the optimization I'm doing here and found a discrepancy with this condition. The behavior changes when the condition is removed, which I didn’t expect.

If `allowed_shipping_countries` is set to the list of all countries provided by WooCommerce, ECE displays a warning in the browser console, indicating that those are invalid country codes:

```
- AS: American Samoa
- PW: Belau
- CX: Christmas Island
- CC: Cocos (Keeling) Islands
- CU: Cuba
- HM: Heard Island and McDonald Islands
- IR: Iran
- MH: Marshall Islands
- FM: Micronesia
- NF: Norfolk Island
- KP: North Korea
- MP: Northern Mariana Islands
- SY: Syria
- UM: United States (US) Minor Outlying Islands
- VI: Virgin Islands (US)
```

Warnings:

```
controller-843927f4a5c13aa00b34dffe4221429a.js:1 [Stripe.js] invalid country code "AS"; "AS" is not a valid 2-letter country code
warn @ controller-843927f4a5c13aa00b34dffe4221429a.js:1
warn @ controller-843927f4a5c13aa00b34dffe4221429a.js:1
(anonymous) @ controller-843927f4a5c13aa00b34dffe4221429a.js:1
controller-843927f4a5c13aa00b34dffe4221429a.js:1 [Stripe.js] invalid country code "PW"; "PW" is not a valid 2-letter country code
warn @ controller-843927f4a5c13aa00b34dffe4221429a.js:1
warn @ controller-843927f4a5c13aa00b34dffe4221429a.js:1
...
```

and the payment sheet doesn't allow selecting a shipping address from those countries. However, when the condition is added and we avoid passing a list, leaving ECE to support all countries, it does allow checkout with addresses from those countries. So far, I’ve tested with AS (American Samoa) and PW (Belau or Palau) and it was able to checkout without problems.

For reference, I haven’t been able to define addresses for some countries like Cuba or Iran. I assume if Google Pay allowed me to add these addresses, ECE would allow me to checkout using a shipping address from those countries.

@bborman22 @pierorocca, what do you think we should do in this case? Should we keep the condition or remove it? If we keep the condition, we address the potential issue [mentioned](https://github.com/Automattic/woocommerce-payments/pull/9592/files#r1807700900) earlier but ECE would allow **shipping** to restricted countries. However, if we remove the condition, we won't resolve the potential issue [mentioned](https://github.com/Automattic/woocommerce-payments/pull/9592/files#r1807700900) earlier, and it won’t be possible to select an address from a country listed above for shipping.

Stripe's behavior seems inconsistent. I'd expect that if any of these country codes are restricted or not allowed, they wouldn't be permitted by default when no countries are passed to `allowedShippingCountries`.

pierorocca replied:
Some of these are restricted countries e.g. Iran, Cuba, Syria, North Korea. Let me ask Stripe about an easier use case like US Virgin Islands.

pierorocca replied:
Response from Stripe:

>Dug into country validation checks and what's supported in Elements...here is what I see,
We skip some countries (such as KP for North Korea and IR for Iran) due to sanctions, along with some regions that are known US or Australian territories, since they are listed as administrative areas within those countries too, and hence not considered as separate
countries.
>
>Similar check applies on Connect as well - here is a [support article](https://support.stripe.com/questions/connect-address-validation#:~:text=Additional%20Information) that has the list.

pierorocca replied:
If you click into the support link, you'll see that Stripe doesn't consider them countries AND for US territories they don't support them either as States that can be validated.

asumaran replied:
I've removed the condition, as it seems the safest option to honor the restrictions. Stripe should address the issue where restricted countries are allowed for shipping when `allowedShippingCountries` is not defined.

asumaran committed:
Remove the condition that enables optimization when sending all country codes

asumaran committed:
Merge branch 'develop' into 9104-add-support-for-allowedshippingcountries-in-express-checkout-element

rafaelzaleski reviewed (approved):
LGTM!

✅ Verified checkout functionality across all specified shipping and selling location configurations (product, cart, and checkout pages—blocks and shortcode). All cases functioned as expected.
